shark/src/main/java/shark/internal/PathFinder.kt

```diff
@@ -462,6 +473,51 @@ internal class PathFinder(
     }
   }
 
+  private fun HeapInstance.readAllNonNullFieldsOfReferenceType(): MutableList<HeapField> {
+    val hprofGraph = graph // DO NOT REMOVE! Getter performs typecast which is expensive!
+    var fieldReader: FieldValuesReader? = null
+    val result = mutableListOf<HeapField>()
+    var skipBytesCount = 0
+    for (heapClass in instanceClass.classHierarchy.toList()) {
+      for (fieldRecord in heapClass.readRecord().fields) {
+        if (fieldRecord.type != PrimitiveType.REFERENCE_HPROF_TYPE) {
+          // Skip all fields that are not references. Track how many bytes to skip
+          skipBytesCount += hprofGraph.getRecordSize(fieldRecord)
+        } else {
+          // Initialize field reader if it's not yet initialized. Replaces `lazy` without synchronization
+          if (fieldReader == null) {
+            fieldReader = hprofGraph.createFieldValuesReader(readRecord())
+          }
+
+          // Skip the accumulated bytes offset
+          fieldReader.skipBytes(skipBytesCount)
+          skipBytesCount = 0
+
+          val fieldValue = ReferenceHolder(fieldReader.readId())
+          if (!fieldValue.isNull) {
+            val fieldName = hprofGraph.fieldName(heapClass.objectId, fieldRecord.nameStringId)
+            result.add(HeapField(heapClass, fieldName, HeapValue(hprofGraph, fieldValue)))
+          }
+        }
+      }
+    }
+    return result
+  }
+
+  private fun HeapGraph.getRecordSize(field: FieldRecord) =
+    when (field.type) {
```

Comment:
Replace all this with `PrimitiveType.byteSizeByHprofType[field.type]`

Reply:
Couple problems with this approach:
1. Since it's a map, using `getValue()` or `[]!!` syntax will both lead to unnecessary boxing of the int value
2. Performance-wise it's not great. It involves searching for a key in the map, while the existing uglier version is a set of if-else branches.

Reply:
Can you replace `PrimitiveType.byteSizeByHprofType` with the faster & uglier implementation?

Reply:
I looked into it but can't figure out how to easily swap it without changing a lot of unrelated to this PR things. The map is used to construct another map, there seem to be some copy&pasted code.
Mind if I do it as a separate PR?

Reply:
btw in my other big refactoring PR I took yet another approach: create an array from the map where each key is the array index. Makes the array slightly bigger but very fast lookups.